### PR TITLE
[PORT] Add getNextViableDate, getPreviousViableDate, getNextViableTime, getPreviousViableTime pre-built functions

### PR DESCRIPTION
**Pull request by cosmicshuai**
fixes: #2559 

Assume today is 2020-06-12 and current time is 15:42:21
getNextViableDate:
description: given a timex property of a valid month and day of month, evaluate the next viable date based on the current time at certain timezone
The signature is getNextViableDate(timex: string, timezone? : string) -> result:string
getNextViableDate("XXXX-12-20", "America/Los_Angelas") -> "2020-12-20"
getNextViableDate("XXXX-02-29") -> "2024-02-29"

getPreviousViableDate:
description: given a timex property of a valid month and day of month, evaluate the previous viable date based on the current time at certain timezone
The signature is getPreviousViableDate(timex: string, timezone? : string) -> result:string
getPreviousViableDate("XXXX-12-20", "Eastern Standard Time") -> "2019-12-20"
getPreviousViableDate("XXXX-02-29") -> "2020-02-29"

getNextViableTime:
description: given a timex property of a valid minutes and seconds, evaluate the next viable time based on the current time at certain timezone
The signature is getNextViableTime(timex: string, timezone? : string) -> result:string
getNextViableTime("TXX:12:14", "Asia/Tokyo") -> "T16:12:14"
getNextViableTime("TXX:52:14) -> "T15:52:14"

getPreviousViableTime:
description: given a timex property of a valid minutes and seconds, evaluate the previous viable time based on the current time at certain timezone
The signature is getPreviousViableTime(timex: string, timezone? : string) -> result:string
getPreviousViableTime("TXX:12:14", 'Europe/London') -> "T15:12:14"
getPreviousViableTime("TXX:52:14") -> "T14:52:14"

### CODE REVIEW

File: libraries/adaptive-expressions/src/expressionEvaluator.ts [hunk 1]
```diff
@@ -14,7 +14,7 @@ import { ReturnType } from './returnType';
  * Delegate for doing static validation on an expression.
  * Validators can and should throw exceptions if the expression is not valid.
  */
-export type ValidateExpressionDelegate = (expression: Expression) => any;
+export type ValidateExpressionDelegate = (expression: any) => any;
```
Comment (chrimc62) [inline libraries/adaptive-expressions/src/expressionEvaluator.ts:17]: >any [](start = 54, length = 3)

Why did all of these change from Expression to any?  It must be an Expression, right?

File: libraries/adaptive-expressions/src/builtinFunctions/getNextViableDate.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+import { MemoryInterface } from '../memory/memoryInterface';
+import { Options } from '../options';
+import { TimeZoneConverter } from '../timeZoneConverter';
+import { tz } from 'moment-timezone';
+import moment from 'moment';
+import {TimexProperty} from '@microsoft/recognizers-text-data-types-timex-expression';
+
+/**
+ * Return true if a given TimexProperty or Timex expression refers to the present.
+ */
```
Comment (chrimc62) [inline libraries/adaptive-expressions/src/builtinFunctions/getNextViableDate.ts:23]: Return true???

File: libraries/adaptive-expressions/src/builtinFunctions/getNextViableTime.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import { FunctionUtils } from '../functionUtils';
+import { MemoryInterface } from '../memory/memoryInterface';
+import { Options } from '../options';
+import { TimeZoneConverter } from '../timeZoneConverter';
+import { tz } from 'moment-timezone';
+import moment from 'moment';
+import {TimexProperty, Time} from '@microsoft/recognizers-text-data-types-timex-expression';
+/**
+ * Return true if a given TimexProperty or Timex expression refers to the present.
+ */
```
Comment (chrimc62) [inline libraries/adaptive-expressions/src/builtinFunctions/getNextViableTime.ts:22]: Return true???

File: libraries/adaptive-expressions/src/builtinFunctions/getPreviousViableDate.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+import { MemoryInterface } from '../memory/memoryInterface';
+import { Options } from '../options';
+import { TimeZoneConverter } from '../timeZoneConverter';
+import { tz } from 'moment-timezone';
+import moment from 'moment';
+import {TimexProperty} from '@microsoft/recognizers-text-data-types-timex-expression';
+
+/**
+ * Return true if a given TimexProperty or Timex expression refers to the present.
+ */
```
Comment (chrimc62) [inline libraries/adaptive-expressions/src/builtinFunctions/getPreviousViableDate.ts:23]: Return true???

File: libraries/adaptive-expressions/src/builtinFunctions/getPreviousViableTime.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+import { FunctionUtils } from '../functionUtils';
+import { MemoryInterface } from '../memory/memoryInterface';
+import { Options } from '../options';
+import { TimeZoneConverter } from '../timeZoneConverter';
+import { tz } from 'moment-timezone';
+import moment from 'moment';
+import {TimexProperty, Time} from '@microsoft/recognizers-text-data-types-timex-expression';
+/**
+ * Return true if a given TimexProperty or Timex expression refers to the present.
+ */
```
Comment (chrimc62) [inline libraries/adaptive-expressions/src/builtinFunctions/getPreviousViableTime.ts:22]: Return true???

File: libraries/adaptive-expressions/src/expressionEvaluator.ts (hunk 1 above)
Comment (cosmicshuai) [reply libraries/adaptive-expressions/src/expressionEvaluator.ts]: right, this should be an Expression. Fixed.

File: libraries/adaptive-expressions/src/builtinFunctions/getNextViableDate.ts (hunk 2 above)
Comment (cosmicshuai) [reply libraries/adaptive-expressions/src/builtinFunctions/getNextViableDate.ts:23]: description modified.

File: libraries/adaptive-expressions/src/builtinFunctions/getPreviousViableDate.ts (hunk 4 above)
Comment (cosmicshuai) [reply libraries/adaptive-expressions/src/builtinFunctions/getPreviousViableDate.ts:23]: description modified.

File: libraries/adaptive-expressions/src/builtinFunctions/getNextViableTime.ts (hunk 3 above)
Comment (cosmicshuai) [reply libraries/adaptive-expressions/src/builtinFunctions/getNextViableTime.ts:22]: description modified.

File: libraries/adaptive-expressions/src/expressionEvaluator.ts [hunk 6]
```diff
@@ -82,5 +82,5 @@ export class ExpressionEvaluator {
      * Validate an expression.
      * @param expression Expression to validate.
      */
-    public validateExpression = (expression: Expression): void => this._validator(expression);
+    public validateExpression = (expression: any): void => this._validator(expression);
```
Comment (chrimc62) [inline libraries/adaptive-expressions/src/expressionEvaluator.ts:85]: >any [](start = 45, length = 3)

Should be Expression

File: libraries/adaptive-expressions/src/expressionEvaluator.ts [hunk 7]
```diff
@@ -69,7 +69,7 @@ export class ExpressionEvaluator {
         this._evaluator = evaluator;
         this.returnType = returnType;
         // eslint-disable-next-line @typescript-eslint/no-unused-vars
-        this._validator = validator || ((expr: Expression): any => { });
+        this._validator = validator || ((expr: any): any => { });
```
Comment (chrimc62) [inline libraries/adaptive-expressions/src/expressionEvaluator.ts:72]: >any [](start = 47, length = 3)

Should be Expression

**Comment by chrimc62** [comment]
Fix the two remaining any -> Expression and you are good to go.

File: libraries/adaptive-expressions/src/expressionEvaluator.ts (hunk 6 above)
Comment (cosmicshuai) [reply libraries/adaptive-expressions/src/expressionEvaluator.ts]: fixed

File: libraries/adaptive-expressions/src/expressionEvaluator.ts (hunk 7 above)
Comment (cosmicshuai) [reply libraries/adaptive-expressions/src/expressionEvaluator.ts]: fixed

**Pull request merged**